modules/sdk-coin-sol/src/sol.ts

```diff
@@ -1043,15 +1041,63 @@ export class Sol extends BaseCoin {
     }
 
     const factory = this.getBuilder();
-    const walletCoin = this.getChain();
 
     let txBuilder;
-    const blockhash = await this.getBlockhash();
+    let blockhash;
+    const recovertTxns: BaseBroadcastTransactionResult[] = [];
 
     const rentExemptAmount = await this.getRentExemptAmount();
 
-    // if this is closeATA recovery
+    // do token recovery before closing ATA address
+    // check if any token is present on the closeAtaAddress
+    const tokenInfo = await this.getTokenAccountInfo(params.closeAtaAddress);
+    const tokenBalance = Number(tokenInfo.info.tokenAmount.amount);
+
+    if (tokenBalance > 0) {
+      // closeATA address has some token balance, it needs to be withdrawn before closing ATA
+      console.log(
+        `closeATA address ${params.closeAtaAddress} has token balance ${tokenBalance}, it needs to be withdrawn before closing ATA address`
+      );
+
+      if (!params.recoveryDestinationAtaAddress || !this.isValidAddress(params.recoveryDestinationAtaAddress)) {
```

Comment:
FYI : recoveryDestinationAtaAddress should be belong to same token account

Reply:
Yes correct, we will provide this detail in coin sandbox script clearly.